url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/engine/impl/ProcessModelToSTS.java

```diff
@@ -0,0 +1,321 @@
+		ProcessNode start = model.getStartNode();
+		return transformProcessModel(model, start);
+	}
+	
+	public STS transformProcessModel(ProcessModel model, ProcessNode start) {
+		init();
+		State initialState = createState();			
+		transform(start, initialState, model);
+		return new DefaultSTS(states, initialState, inputActions, 
+				outputActions, transitions);	
+	}
+	
+	private void init() {
+		this.stateIndex = 0;
+		this.states = new HashSet<State>();
+		xorJoinState = new HashMap<XorJoin, State>();
+		
+		inputActions = new HashSet<Action>();
+		outputActions = new HashSet<Action>();
+		transitions = new HashSet<Transition>();
+	}
+		
+	private State createState() {
+		return createState("s" + stateIndex);
+	}
+	
+	private State createState(String name) {
+		State newState = new DefaultState(name);
+		stateIndex++;
+		states.add(newState);
+		return newState;
+	}
+	
+	private void transform(ProcessNode node, State fromState, ProcessModel model) {
+		if (node instanceof ActivityNode) {
+			transformActivityNode((ActivityNode) node, fromState, model);
+		} 
+		if (node instanceof XorSplit) {
+			transformXorSplit((XorSplit) node, fromState, model);
+		}
+		if (node instanceof XorJoin) {
+			transformXorJoin((XorJoin) node, fromState, model);
+		}
+		if (node instanceof AndSplit) {
+			transformAndSplit((AndSplit) node, fromState, model);
+		}
+	}
+	
+	private void transformActivityNode(ActivityNode actNode, State fromState, ProcessModel model) {
+		State nextState = transformActivity(actNode, fromState, model);
+		for (ProcessEdge edge : model.outgoingEdgesOf(actNode)) {
+			transform(model.getEdgeTarget(edge), nextState, model);
+		}
+	}
+	
+	private void transformXorSplit(XorSplit splitNode, State fromState, ProcessModel model) {
+		State nextState = createState();
+		
+		xorCount++;
+		Action action = new DefaultAction("Xor" + xorCount, true, model);
+		Transition trans = addActionToSTS(action, fromState, nextState, new StateFormula());
+		int i = 1;
+		//TODO if a branch has no condition, negate and conjoin the conditions on the other branches
+		
+		Map<ProcessNode, State> nextNodes = new HashMap<ProcessNode, State>();
+		for (ProcessEdge edge : model.outgoingEdgesOf(splitNode)) {
+			State branchState = createState();
+			Action caseAction = new DefaultAction("Case" + xorCount + i, false, model);
+			i++;
+			StateFormula condition = edge.getCondition();
+			if (condition.isEmpty()) {
+				for (ProcessEdge otherEdge : model.outgoingEdgesOf(splitNode)) {
+					if (!otherEdge.equals(edge) && !otherEdge.getCondition().isEmpty()) {
+						condition.add(otherEdge.getCondition().getNegation());
+					}
+				}
+			}
+			addActionToSTS(caseAction, nextState, branchState, condition);
+			
+			logger.trace("Adding sts action for XOR edge with condition " + condition);
+			nextNodes.put(model.getEdgeTarget(edge), branchState);
+		}
+
+		for (ProcessNode nextNode : nextNodes.keySet()) {
+			transform(nextNode, nextNodes.get(nextNode), model);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.